### PR TITLE
Initial stacks API

### PR DESCRIPTION
**Pull request by jnummelin**
This PR brings the stack concept to the server side.

### CODE REVIEW

**Comment by jnummelin** [comment]
Fixes #702

File: server/app/models/stack.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  include Mongoid::Enum
+
+  field :name, type: String
+  field :version, type: String
+  enum :state, [:initialized, :deployed, :terminated]
+
+  belongs_to :grid
+
+  has_many :grid_services
+  #has_many :audit_logs
```
Comment (jakolehm) [inline server/app/models/stack.rb]: Let's remove these commented code blocks... if we don't need them then they should not be here :)

File: server/app/models/stack.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  index({ grid_id: 1 })
+  index({ name: 1 })
+
+  validates_presence_of :name
+  #validates_uniqueness_of :name, scope: [:grid_id]
```
Comment (jakolehm) [inline server/app/models/stack.rb]: Probably we need to validate uniqueness?

File: server/app/mutations/grid_services/add_env.rb [hunk 3]
```diff
@@ -7,6 +7,7 @@ class AddEnv < Mutations::Command
 
     def execute
       self.grid_service.env << env
+      grid_service.revision += 1
```
Comment (jakolehm) [inline server/app/mutations/grid_services/add_env.rb]: Can revision be nil? Should we cast revision to integer first?

Comment (jnummelin) [inline server/app/mutations/grid_services/add_env.rb]: It shouldn't be nil as the model defaults to 1

File: server/app/mutations/stacks/delete.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        GridServices::Delete.validate(current_user: self.current_user, grid_service: service)
+      end
+    end
+
+    def execute
+      self.stack.state = :terminated
+      self.stack.save
+      # Remove all services of the stack
+      self.stack.grid_services.each do |service|
+        worker(:grid_service_remove).async.perform(service.id)
```
Comment (jakolehm) [inline server/app/mutations/stacks/delete.rb]: Why no mutation here?

File: server/app/routes/v1/stacks_api.rb [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+      def update_stack(stack, data)
+        data[:grid] = @grid
+        data[:current_user] = current_user
+        data[:stack] = stack
+        outcome = Stacks::Update.run(data)
+
+        if outcome.success?
+          @stack = outcome.result
+          #audit_event(request, @grid, @grid_secret, 'update', nil, [:body])
```
Comment (jakolehm) [inline server/app/routes/v1/stacks_api.rb]: I would like to have audit log :)

File: server/app/models/grid.rb [hunk 6]
```diff
@@ -28,6 +28,7 @@ def self.default_overlay_cidr
   has_many :audit_logs
   has_many :registries, dependent: :delete
   has_many :overlay_cidrs, dependent: :delete
+  has_many :stacks
```
Comment (jakolehm) [inline server/app/models/grid.rb]: Should this have `dependent: :destroy`?

File: server/app/models/stack.rb (hunk 2 above)
Comment (jnummelin) [inline server/app/models/stack.rb]: Yes, and in the scope of grid.

File: server/app/routes/v1/stacks_api.rb (hunk 5 above)
Comment (jnummelin) [inline server/app/routes/v1/stacks_api.rb]: You shall have one. :D

File: server/app/models/grid.rb (hunk 6 above)
Comment (jnummelin) [inline server/app/models/grid.rb]: yep

**Comment by jnummelin** [comment]
Comments addressed in last commit

File: server/app/models/stack.rb (hunk 1 above)
Comment (jakolehm) [inline server/app/models/stack.rb]: How about these commented out lines...?

Comment (jnummelin) [inline server/app/models/stack.rb]: removed in last push

**Comment by jakolehm** [comment]
LGTM!

**Pull request merged**